sklearn/ensemble/_hist_gradient_boosting/tests/test_gradient_boosting.py

```diff
@@ -172,3 +193,242 @@ def test_binning_train_validation_are_separated():
+    (HistGradientBoostingRegressor, X_regression, y_regression)
+])
+def test_warm_start_max_depth(GradientBoosting, X, y):
+    # Test if possible to fit trees of different depth in ensemble.
+    gb = GradientBoosting(max_iter=100, min_samples_leaf=1,
+                          warm_start=True, max_depth=2)
+    gb.fit(X, y)
+    gb.set_params(max_iter=110, max_depth=3)
+    gb.fit(X, y)
+
+    # First 100 trees have max_depth == 2
+    for i in range(100):
+        assert_equal(gb._predictors[i][0].get_max_depth(), 2)
+    # Last 10 trees have max_depth == 3
+    for i in range(1, 11):
+        assert_equal(gb._predictors[-i][0].get_max_depth(), 3)
+
+
+@pytest.mark.parametrize('GradientBoosting, X, y', [
+    (HistGradientBoostingClassifier, X_classification, y_classification),
+    (HistGradientBoostingRegressor, X_regression, y_regression)
+])
+def test_warm_start_early_stopping(GradientBoosting, X, y):
+    # Make sure that early stopping occurs after a small number of iterations
+    # when fitting a second time with warm starting.
+
+    n_iter_no_change = 5
+    gb = GradientBoosting(
+        n_iter_no_change=n_iter_no_change, max_iter=10000,
+        random_state=42, warm_start=True, tol=1e-3
+    )
+    gb.fit(X, y)
+    n_iter_first_fit = gb.n_iter_
+    gb.fit(X, y)
+    n_iter_second_fit = gb.n_iter_
+    assert n_iter_second_fit - n_iter_first_fit < n_iter_no_change
+
+
+@pytest.mark.parametrize('GradientBoosting, X, y', [
+    (HistGradientBoostingClassifier, X_classification, y_classification),
+    (HistGradientBoostingRegressor, X_regression, y_regression)
+])
+def test_warm_start_equal_n_estimators(GradientBoosting, X, y):
+    # Test if warm start with equal n_estimators does nothing
+    gb_1 = GradientBoosting(max_depth=2)
+    gb_1.fit(X, y)
+
+    gb_2 = clone(gb_1)
+    gb_2.set_params(max_iter=gb_1.max_iter, warm_start=True)
+    gb_2.fit(X, y)
+
+    # Check that both predictors are equal
+    _assert_predictor_equal(gb_1, gb_2, X)
+
+
+@pytest.mark.parametrize('GradientBoosting, X, y', [
+    (HistGradientBoostingClassifier, X_classification, y_classification),
+    (HistGradientBoostingRegressor, X_regression, y_regression)
+])
+def test_warm_start_clear(GradientBoosting, X, y):
+    # Test if fit clears state.
+    gb_1 = GradientBoosting(max_depth=2)
+    gb_1.fit(X, y)
+
+    gb_2 = GradientBoosting(max_depth=2, warm_start=True)
+    gb_2.fit(X, y)  # inits state
+    gb_2.set_params(warm_start=False)
+    gb_2.fit(X, y)  # clears old state and equals est
+
+    # Check that both predictors are equal
+    _assert_predictor_equal(gb_1, gb_2, X)
+
+
+@pytest.mark.parametrize('GradientBoosting, X, y', [
+    (HistGradientBoostingClassifier, X_classification, y_classification),
+    (HistGradientBoostingRegressor, X_regression, y_regression)
+])
+def test_identical_train_val_split_int(GradientBoosting, X, y):
+    # Test if identical splits are generated when random_state is an int.
```

Comment:
I think we could factorize all the tests below with the following one:

```python
@pytest.mark.parametrize('GradientBoosting, X, y', [
    (HistGradientBoostingClassifier, X_classification, y_classification),
    (HistGradientBoostingRegressor, X_regression, y_regression)
])
@pytest.mark.parametrize('rng_type', ('int', 'instance', None))
def test_random_seeds_warm_start(GradientBoosting, X, y, rng_type):
    # Make sure the seeds for train/val split and small trainset subsampling
    # are correctly set in a warm start context.
    def _get_rng(rng_type):
        # Helper to avoid consuming rngs
        if rng_type == 'int':
            return 42
        elif rng_type == 'instance':
            return np.random.RandomState(0)
        else:
            return None

    random_state = _get_rng(rng_type)
    gb_1 = GradientBoosting(n_iter_no_change=5, random_state=random_state)
    gb_1.fit(X, y)
    train_val_seed_1 = gb_1._train_val_split_seed
    small_trainset_seed_1 = gb_1._small_trainset_seed

    random_state = _get_rng(rng_type)
    gb_2 = GradientBoosting(n_iter_no_change=5, random_state=random_state,
                            warm_start=True)
    gb_2.fit(X, y)  # inits state
    train_val_seed_2 = gb_2._train_val_split_seed
    small_trainset_seed_2 = gb_2._small_trainset_seed
    gb_2.fit(X, y)  # clears old state and equals est
    train_val_seed_3 = gb_2._train_val_split_seed
    small_trainset_seed_3 = gb_2._small_trainset_seed

    # Check that all seeds are equal
    if rng_type is None:
        assert train_val_seed_1 != train_val_seed_2
        assert small_trainset_seed_1 != small_trainset_seed_2
    else:
        assert train_val_seed_1 == train_val_seed_2
        assert small_trainset_seed_1 == small_trainset_seed_2

    assert train_val_seed_2 == train_val_seed_3
    assert small_trainset_seed_2 == small_trainset_seed_3
```

Reply:
Great idea!